waiter/src/waiter/token.clj

```diff
@@ -170,17 +206,23 @@
           (if (and service-description-template (not-empty service-description-template))
             (let [token-owner (get token-metadata "owner")]
               (if hard-delete
-                (when-not (authz/administer-token? entitlement-manager authenticated-user token token-metadata)
-                  (throw (ex-info "Cannot hard-delete token"
-                                  {:metadata token-metadata
-                                   :status 403
-                                   :user authenticated-user})))
+                (do
+                  (when-not (parse-last-modified headers "if-unmodified-since")
+                    (throw (ex-info "Must specify if-unmodified-since header for token hard deletes"
+                                    {:request-headers headers, :status 400})))
+                  (when-not (authz/administer-token? entitlement-manager authenticated-user token token-metadata)
+                    (throw (ex-info "Cannot hard-delete token"
+                                    {:metadata token-metadata
+                                     :status 403
+                                     :user authenticated-user}))))
                 (when-not (authz/manage-token? entitlement-manager authenticated-user token token-metadata)
                   (throw (ex-info "User not allowed to delete token"
                                   {:owner token-owner
                                    :status 403
                                    :user authenticated-user}))))
-              (delete-service-description-for-token clock synchronize-fn kv-store token token-owner :hard-delete hard-delete)
+              (delete-service-description-for-token clock synchronize-fn kv-store token token-owner
+                                                    :hard-delete hard-delete
+                                                    :if-unmodified-since (compute-last-modified clock headers))
```

Comment:
`delete-service-description-for-token` should take a `long`, but it should be parsed from the header which conforms to the time spec.

Reply:
Done.